RFC-0036-hardware-accelerators-pytorch.org.md

```diff
@@ -0,0 +1,87 @@
+The main page of pytorch.org only shows PyTorch Get-Started  on CUDA, ROCm and CPU, but the hardware accelerator support is more diverse and the ecosystem is rapidly evolving with new accelerators like Intel Gaudi, Huawei Ascend, Google TPU, etc.
+The proposal in this RFC is to enable the visibility of broad set of accelerators that supports PyTorch for the community, and create a space for hardware vendors to provide instructions and details about their PyTorch support.  
+
+
+## **Motivation**
+There are 2 primary motivations behind this RFC:
+1. Evangelize PyTorch as an open source framework that supports a wide variety of hardware.
+   * PyTorch is supported on more compute platforms than CUDA, ROCm and CPU .
+   * PyTorch can maintain and enhance its competitive position in the machine learning frameworks  by enabling the visibility of rich set of hardware accelerators on which PyTorch can run.
+2. Provide a consolidated entry point where to look for more information about supported hardware accelerators.
+   * pytorch.org is a starting point for many users. The page can help users to find more details.
+
+
+
+## **Proposed Implementation**
+The proposal is to enhance pytorch.org by adding 3  new sections:
+* Add new compute platforms to “Install PyTorch” section at the main page of pytorch.org.
+* Add new compute platforms to https://pytorch.org/get-started/.
+* Add new “Compute Platforms” section at the main page of pytorch.org.
+
+### PyTorch integration types
+There are at least 2 ways how compute platforms are integrated with PyTorch:
+1. In-tree – CPU, CUDA, ROCm are developed, built and tested in PyTorch environment. PyTorch is ensuring quality criteria. This approach is limited to only a few compute platforms and it does not scale with number of compute platforms.
+2. Out-of-tree – Integration of other compute platforms like Intel Gaudi or Huawei Ascend is done via additional python package (extension) that needs to be installed on top of PyTorch CPU package. Development, built and testing is done outside of PyTorch. In this case:
+   * PyTorch is ensuring quality of PyTorch CPU package.
+   * Hardware provider is ensuring quality of its extension against PyTorch CPU package.
+
+### Out-of-tree integration vs. device key
+Out-of-tree integration refers to compute platforms integrated with its own device key or PrivateUse1.
+Example of out-of-tree integration with its own device key is Intel Gaudi with "HPU" key.
+Example of out-of-tree integration using PrivateUse1 is Huawei Ascend.
+
+
+### Stable vs. nightly builds
+1. Stable builds – the hardware provider shall provide the installation commands when a compute platform is tested against given PyTorch version and it meets the quality criteria set by PyTorch Foundation.    
+2. Preview (nightly) -  Similar to the stable builds, but the hardware provider must implement method to provide quickly fixes for PyTorch nightly.   It should be optional.
+
+### Quality criteria
```

Comment:
I think we want to go a bit deeper into the details of what are going to be the criterions and expectations here.
I am also working on this and I think we need to properly define a couple of "tiers" for backends, defining expectations in details. I think that we can then refine the way we present them from this proposal where top-tier backends will be in the main get-started page, all others will be in the Compute Platforms page linked from get-started.

I'll take a stab at that proposal and come back to you here.

Reply:
@albanD waiting for the proposal from your side.
Please let me know if I can help here.

Reply:
My high level thoughts about quality criteria:

1. There are many types of devices and there should be a way for them to pass quality criteria
- some of them are specialized i.e. inference only, machine learning or more general purpose like GPUs or CPUs
- they differ in capabilities i.e. data types support, distributed ops support
- there should be a way to mark relevant tests for a given device through test marking or more extensive capability reporting by a device
2. PyTorch has lots of unit tests, but they are very often specialized to CUDA
- it will take a few PyTorch releases to generalize tests for other devices i.e. we have number of generic changes that will enable Intel Gaudi and other devices with PT UTs. We will follow up with RFCs and UT PRs 
- PyTorch UTs are probably the best ones to verify quality of PyTorch support by a device
3. The quality criteria should be set to reasonable level until PyTorch UT changes are upstreamed i.e. initially start with 50% pass rate for PT 2.4 and gradually increase it in the next releases
4. I agree that there should be a couple of "tiers"
- Do you expect that 1st tier out of tree backends (the ones from the main get-started page) will work with PyTorch CPU from pytorch.org?

Reply:
@albanD any update?

Reply:
@albanD are there any details that can be shared with us?

Reply:
Hey!
Sorry I have been held out in other projects (PT conference right now) and was not able to look at this yet.